cpp/src/arrow/array/data.h

```diff
@@ -434,16 +455,36 @@ struct ARROW_EXPORT ArraySpan {
   inline bool IsValid(int64_t i) const {
     if (this->buffers[0].data != NULLPTR) {
       return bit_util::GetBit(this->buffers[0].data, i + this->offset);
+    }
+    const auto type = this->type->id();
+    if (type == Type::SPARSE_UNION) {
+      return !IsNullSparseUnion(i);
+    } else if (type == Type::DENSE_UNION) {
+      return !IsNullDenseUnion(i);
+    } else if (type == Type::RUN_END_ENCODED) {
+      return !IsNullRunEndEncoded(i);
+    }
+    return this->null_count != this->length;
+  }
+
+  template <typename ArrowType>
+  inline bool IsNullFast(int64_t i) const {
+    return !IsValidFast<ArrowType>(i);
+  }
+
+  template <typename ArrowType>
+  inline bool IsValidFast(int64_t i) const {
+    if constexpr (ArrowType::type_id == Type::NA) {
```

Comment:
Do we need to add a DCHECK to check that `ArrowType` is equal to the real type in the Array?

Reply:
I was tempted, but it adds too high of an overhead for this function as it's used from tight loops and is expected to be fully inlined. Even considering that it would be affecting only debug builds.

Reply:
Maybe we just need care the release performance? An array using outerside `ArrowType` without checking is so dangerous...

Reply:
Okay, though it's still a bit wired for me. I think we can hear about others idea.

Reply:
To be more clear: the body of these member functions in the `.cc` file, where `logging.h` can be used, contains runtime checks.

```cpp
    } else if constexpr (ArrowType::type_id == Type::SPARSE_UNION) {
      return !IsNullSparseUnion(i);
    } else if constexpr (ArrowType::type_id == Type::DENSE_UNION) {
      return !IsNullDenseUnion(i);
    } else if constexpr (ArrowType::type_id == Type::RUN_END_ENCODED) {
      return !IsNullRunEndEncoded(i);
   ```